add kubebuilder/operator-sdk FAQ (#4411)
operator-framework/operator-sdk: website/content/en/docs/faqs/_index.md

weight: 10
---

> ## What's the different between kubebuilder and operator-sdk?

Reviewer: ```suggestion
## What are the differences between Kubebuilder and Operator-SDK?
```

> Currently, kubebuilder is used as a plugin for many of the operator-sdk features around golang-based operators. Thus, much of the scaffolding operator-sdk generates for a golang operator should look familiar if you are familiar with kubebuilder. See [this post] for further discussion of how kubebuilder is used within operator-sdk. Note that pre-1.0.0 versions of operator-sdk did not use kubebuilder, although they still used controller-runtime to generate scaffolding. See [this issue](https://github.com/operator-framework/operator-sdk/issues/1758) for past discussion of legacy differences.
> 

Reviewer: ```suggestion
SDK and Kubebuilder are integrated projects which have the purpose to help you to build your operator projects. Both tools leverage on [controller-runtime][controller-runtime] and will follow the same standard layout. For further information see the doc [What’s in a basic project?][kb-doc-what-is-a-basic-project]. 

However, SDK provides features to help you with some actions such as; to integrate your project with [OLM][olm] and to distribute your projects on [OperatorHub.io][operatorhub.io] as the OCP catalogue, [Scorecard][scorecard-doc] to help you test your projects. Nonetheless, SDK also allows you to work with other languages such as Ansible and Helm to create operators as well. 

Also, to have a better context over the advantages to use SDK as its differences over Kubebuilder, check the [comment][issue-comment] and the blog post [Operator SDK Reaches v1.0][operator-sdk-reaches-v1.0]. 

```

Reviewer: Following the links:

[controller-runtime]: https://github.com/kubernetes-sigs/controller-runtime
[kb-doc-what-is-a-basic-project]: https://book.kubebuilder.io/cronjob-tutorial/basic-project.html
[olm]:  https://github.com/operator-framework/operator-lifecycle-manager
[operatorhub.io]: https://operatorhub.io/
[scorecard-doc]: https://sdk.operatorframework.io/docs/advanced-topics/scorecard/
[operator-sdk-reaches-v1.0]:

weight: 10
---

## What are the the differences between Kubebuilder and Operator-SDK?

> Kubebuilder and Operator-SDK are integrated projects which both allow you to quickly scaffold an operator project. Both tools make use [controller-runtime][controller-runtime] and the resulting projects will have the same basic layout. For more information see the doc about [what’s in a basic project][kb-doc-what-is-a-basic-project].  

Reviewer 2: ```suggestion
Kubebuilder and Operator SDK are both projects that allow you to quickly create and manage an operator project. Operator SDK uses kubebuilder under the hood to do so for Go projects, such that the `operator-sdk` CLI tool will work with a project created by `kubebuilder`. Therefore each project makes use of [controller-runtime][controller-runtime] and will have the same [basic layout][kb-doc-what-is-a-basic-project].
```

Author: what is the definitive stylization of the name of our project? I've seen Operator-SDK, operator-sdk, and now Operator SDK

Reviewer 2: Its either all lowercase and hyphenated `operator-sdk` (when referring to the binary itself or repo name) or title'd and not hyphenated `Operator SDK` (when formally referring to the Operator SDK project). I personally prefer always using `operator-sdk` to avoid ambiguity.

> Operator-SDK offers additional features on top of the basic project scaffolding that Kubebuilder provides. It allows you to generate scaffolding to integrate your project with the [Operator Lifecycle Manager(OLM)][olm], a tool for managing the deployment, use, and upgrade of an operator on an actual in-use cluster. This scaffolding can also be used to distribute your operator on [OperatorHub][operatorhub.io], a community hub for publishing operators. Operator-SDK also offers the capability to test your deployed operator with [Scorecard][scorecard-doc], as well as the supporting other languages such as Ansible and Helm to create operators.

Reviewer 2: ```suggestion
Operator SDK offers additional features on top of the basic project scaffolding that Kubebuilder provides. By default, `operator-sdk init` generates a project integrated with:
- [Operator Lifecycle Manager][olm], an installation and runtime management system for operators
- [OperatorHub][operatorhub.io], a community hub for publishing operators
- Operator SDK [scorecard][scorecard-doc], a tool for ensuring operator best-practices and developing cluster tests
Operator SDK supports operator types other than Go as well, such as Ansible and Helm.
```

> Operator-SDK offers additional features on top of the basic project scaffolding that Kubebuilder provides. It allows you to generate scaffolding to integrate your project with the [Operator Lifecycle Manager][olm] or OLM, a tool for mananaging the deployment, use, and upgrade of an operator on an actual in-use cluster. This scaffolding can also be used to distribute your operator on [OperatorHub][operatorhub.io], a community hub for publishing operators. Operator-SDK also has support for [Scorecard][scorecard-doc], a set of features that allow you to scaffold and run a suite of integration tests for a deployed operator. Operator-SDK also allows you to work with other languages such as Ansible and Helm to create operators as well.

Reviewer: ```suggestion
Operator-SDK offers additional features on top of the basic project scaffolding that Kubebuilder provides. It allows you to generate scaffolding to integrate your project with the [OLM (Operator Lifecycle Manager)][olm] and to distribute your operator on [OperatorHub][operatorhub.io]. Operator-SDK also offers features to test your project such as [Scorecard][scorecard-doc] and then, by using Operator-SDK you also are allowed to work with other languages such as Ansible and Helm to create operators as well.
```

> For further context about the relationship between Kubebuilder and Operator-SDK, see [this blog post][operator-sdk-reaches-v1.0]. Note that in the past, the relationship was somewhat different, see [this comment][legacy-issue] about some of the legacy functionality from pre-1.0.0 versions of Operator-SDK.

Reviewer 2: There's no need to reference legacy differences. It might confuse readers.
```suggestion
For further context about the relationship between Kubebuilder and Operator SDK, see [this blog post][operator-sdk-reaches-v1.0].
```

> For further context about the relationship between Kubebuilder and Operator-SDK, see [this blog post][operator-sdk-reaches-v1.0]. Note that in the past, the relationship was somewhat different, see [this comment][legacy-issue] about some of the legacy functionality from pre-1.0.0 versions of OperatorSDK.

Reviewer: ```suggestion
For further context about the relationship between Kubebuilder and Operator-SDK, see [this blog post][operator-sdk-reaches-v1.0] and [this comment][legacy-issue].
```

Author: I'm not so sure about this change. I want to mention that things were different pre 1.0.0

Reviewer: Hi @jberkhahn, 
 
the issue explain the diffs and the blog teh diffs + what was achieved with 1.0

Now run `make manifests` to update your `role.yaml`.

[client.Reader]:https://godoc.org/sigs.k8s.io/controller-runtime/pkg/client#Reader
[controller-runtime]: https://github.com/kubernetes-sigs/controller-runtime
[cr-faq]:https://github.com/kubernetes-sigs/controller-runtime/blob/master/FAQ.md
[finalizer]:/docs/building-operators/golang/advanced-topics/#handle-cleanup-on-deletion
[kb-doc-what-is-a-basic-project]: https://book.kubebuilder.io/cronjob-tutorial/basic-project.html
[kube-apiserver_options]: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/#options

> [legacy-issue]: https://github.com/operator-framework/operator-sdk/issues/1758#issuecomment-517432349

Reviewer 2: ```suggestion
```

Reviewer 2: @jberkhahn last one then g2g.

Author: fixed.
